substrate/client/consensus/beefy/src/communication/gossip.rs

```diff
@@ -385,9 +420,28 @@ where
 		mut data: &[u8],
 	) -> ValidationResult<B::Hash> {
 		let raw = data;
+		// Have fisherman check the vote or proof for fork equivocations regardless of
+		// the action on it:
+		// 1. We check votes/proofs on past rounds and active rounds since they might equivocate
+		//    against grandpa-finalized state.
+		// 2. We check votes/proofs on future non-active rounds since these should not have been
+		//    voted on yet (from the client's perspective). In case the block is not even in the
+		//    client's history, but is in fact already finalized, the resulting equivocation report
+		//    will be ineffectual.
+		// The check is best-effort and ignores errors such as state pruned. Accepted votes are also
+		// checked against vote equivocations in
+		// `sc_consensus_beefy::worker::BeefyWorker::handle_vote`
 		let action = match GossipMessage::<B>::decode_all(&mut data) {
-			Ok(GossipMessage::Vote(msg)) => self.validate_vote(msg, sender),
-			Ok(GossipMessage::FinalityProof(proof)) => self.validate_finality_proof(proof, sender),
+			Ok(GossipMessage::Vote(msg)) => {
+				let action = self.validate_vote(&msg, sender);
+				let _ = self.fisherman.check_vote(msg);
+				action
+			},
+			Ok(GossipMessage::FinalityProof(proof)) => {
+				let action = self.validate_finality_proof(&proof, sender);
+				let _ = self.fisherman.check_proof(proof);
```

Comment:
Here the signatures verification process takes place 2-3 times:
- once for `self.validate_finality_proof(&proof, sender);`
- 1-2 times for `self.fisherman.check_proof(proof);`, depending on the branch. For example: 
  - once here:
  ```
				if crate::justification::verify_with_validator_set::<B>(
					commitment.block_number,
					&validator_set,
					&BeefyVersionedFinalityProof::<B>::V1(signed_commitment),
				)
  ```
  - and once more here:
  ```
				// report every signer of the bad justification
				let signatories: Vec<_> = validator_set
					.validators()
					.iter()
					.cloned()
					.zip(signatures.into_iter())
					.filter_map(|(id, signature)| match signature {
						Some(sig) => {
							if sp_consensus_beefy::check_commitment_signature::<
								_,
								_,
								BeefySignatureHasher,
							>(&commitment, &id, &sig)
							{
								Some((id, sig.clone()))
							} else {
								None
							}
						},
						None => None,
					})
					.collect();
  ```

I would try to do this more efficiently